timemachine/md/minimizer.py

```diff
@@ -277,3 +278,83 @@ def equilibrate_host(
     ctxt.multiple_steps(np.linspace(0.0, 0.0, n_steps))
 
     return ctxt.get_x_t(), ctxt.get_box()
+
+
+def get_val_and_grad_fn(impls, box, lamb):
+    """
+    Convert impls, box, lamb into a function that only takes in coords.
+    """
+
+    def val_and_grad_fn(coords):
+        nrgs = []
+        grads = []
+        for impl in impls:
+            g, _, u = impl.execute(coords, box, lamb)
+            nrgs.append(u)
+            grads.append(g)
+        return np.sum(nrgs, axis=0), np.sum(grads, axis=0)
+
+    return val_and_grad_fn
+
+
+def local_minimize(x0, val_and_grad_fn, local_idxs):
+    """
+    Minimize a local region given selected idxs.
+
+    Parameters:
+    -----------
+    x0: np.array (N,3)
+        Coordinates
+
+    box0: np.array (3,3)
+        Frozen box
+
+    val_and_grad_fn: f: R^(Nx3,3x3) -> (R^1, R^Nx3)
+        Energy function
+
+    local_idxs: list of int
+        Unique idxs we allow to move.
+
+    Returns
+    -------
+    Optimized set of coordinates.
+
+    """
+
+    assert len(local_idxs) == len(set(local_idxs))
+
+    x_local_shape = (len(local_idxs), 3)
+    u_0, _ = val_and_grad_fn(x0)
+
+    # deal with overflow
+    guard_threshold = 1e6
+
+    def val_and_grad_fn_local(x_local):
+        x_prime = x0.copy()
+        x_prime[local_idxs] = x_local
+        u_full, grad_full = val_and_grad_fn(x_prime)
+        # we minimized the energy too much, likely an overflow
```

Comment:
"we minimized the energy too much" is not quite the right description of what we were seeing, as this issue was encountered on the first step of line search (and all subsequent configurations tried) on one of four test systems.

```suggestion
        # avoid being trapped when overflows spuriously appear as large negative numbers
        # remove after resolution of https://github.com/proteneer/timemachine/issues/481
```